libtcmu_log.c

```diff
@@ -193,6 +211,9 @@ log_internal(int pri, struct tcmu_device *dev, const char *funcname,
 
 	vsnprintf(msg + n, LOG_MSG_LEN - n, fmt, args);
 
+	memcpy(buf, msg, LOG_MSG_LEN);
+	log_output(pri, buf, true);
```

Comment:
I think you need to give me more info o this, because I'm not sure I got the bug.

It looks like we could log the message here, but then below we queue it to be logged again later for the same output type.

Also why do we want to do this for file and stdout outputs and also why do we need to always log both of them from the calling context?

Reply:
Here we just set the file and stdout as bypassing the log RingBuffer, and each log outputing will output the messages to the log file and stdout directly only. 

But only for the syslog, we need the RingBuffer to cache the log messages and avoid the tcmu-runner daemon hangs. As we discussed  before when the syslogd daemon is died, then all the syslog callout will be hanged forever.

The bug we encountered is that when we cache all the log messages in RingBuffer first, and then flush it to all the output, when crash or other cases the very important log messages in RingBuffer may be lost.

After this patch we can make sure that at least we can find some important log messages in the log file.

Reply:
I am going to merge it.

I am just a little worried about adding something that we cannot control in the main IO/error paths. I was trying to actually use things like mlock in these paths so they are more predictable. For example, to log a message the kernel might have to now write out pages to get memory for the write operation. That could hit other issues like slow disks we have to then block on just to write out a message.  This could happen for hundreds of IOs. I think we might hit issues with failover/failure timers on the initiator side if it has to then wait for extra issues like these, but I am just not sure that it is a issue and I think I am just being too paranoid.

Reply:
Yes, while mostly this should be rare to happen(?) And the same time in product cases, the INFO(as default) log level should be used.

IMHO, currently most machines the slow disks shouldn't be a problem, if it happens this could be that the system memories are almost exhausted and it is swapping memories to disk. If so, not only the tcmu-runner daemon will be slowed down but also all the other tasks in the machine.

Did I understand your thoughts correctly ?